pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
 		}
-	}
-	// cleanup target groups
-	return deleteTargetGroups(vpc, awsSession, logger)
-}
-
-// filterTargetGroupsByVPC will find all the target groups in the provided list that are under the provided VPC
-func filterTargetGroupsByVPC(tgs []*elbv2.TargetGroup, vpc *ec2.Vpc, logger logrus.FieldLogger) []*elbv2.TargetGroup {
-	filteredTGs := []*elbv2.TargetGroup{}
 
-	for _, tg := range tgs {
-		if *tg.VpcId == *vpc.VpcId {
-			filteredTGs = append(filteredTGs, tg)
-		}
+		break
 	}
 
-	return filteredTGs
+	return "", nil
 }
 
-func deleteTargetGroups(vpc *ec2.Vpc, awsSession *session.Session, logger logrus.FieldLogger) bool {
-	logger.Debugf("Deleting target groups (%s)", *vpc.VpcId)
-	defer logger.Debugf("Exiting deleting target groups (%s)", *vpc.VpcId)
-	elbv2Client := elbv2.New(awsSession)
-
-	total := 0
-	filteredTGs := []*elbv2.TargetGroup{}
-	if err := elbv2Client.DescribeTargetGroupsPages(&elbv2.DescribeTargetGroupsInput{}, func(results *elbv2.DescribeTargetGroupsOutput, lastPage bool) bool {
-		total += len(results.TargetGroups)
-		filteredTGs = append(filteredTGs, filterTargetGroupsByVPC(results.TargetGroups, vpc, logger)...)
-		return lastPage
-	}); err != nil {
-		logger.Errorf("Error listing target groups: %v", err)
-		return false
-	}
-	logger.Debugf("from %d total target groups, %d scheduled for deletion", total, len(filteredTGs))
+func deleteARN(session *session.Session, arnString string, logger logrus.FieldLogger) error {
+	logger = logger.WithField("arn", arnString)
 
-	if len(filteredTGs) == 0 {
-		// no items left to delete
-		return true
+	parsed, err := arn.Parse(arnString)
+	if err != nil {
+		return err
 	}
 
-	for _, tg := range filteredTGs {
-		logger.Debugf("Deleting target groups: %v", *tg.TargetGroupName)
-		_, err := elbv2Client.DeleteTargetGroup(&elbv2.DeleteTargetGroupInput{
-			TargetGroupArn: tg.TargetGroupArn,
-		})
-		if err != nil {
-			logger.Debugf("Error deleting target groups %v: %v", *tg.TargetGroupName, err)
-		} else {
-			logger.WithField("name", *tg.TargetGroupName).Info("Deleted target group")
-		}
+	switch parsed.Service {
+	case "ec2":
+		return deleteEC2(session, parsed, logger)
+	case "elasticloadbalancing":
+		return deleteElasticLoadBalancing(session, parsed, logger)
+	case "iam":
+		return deleteIAM(session, parsed, logger)
+	case "route53":
+		return deleteRoute53(session, parsed, logger)
+	case "s3":
+		return deleteS3(session, parsed, logger)
+	default:
+		return errors.Errorf("unrecognized ARN service %s (%s)", parsed.Service, arnString)
 	}
-	return false
 }
 
-// rtHasMainAssociation will check whether a given route table has an association marked 'Main'
-func rtHasMainAssociation(rt *ec2.RouteTable) bool {
-	for _, association := range rt.Associations {
-		if *association.Main == true {
-			return true
-		}
+func deleteEC2(session *session.Session, arn arn.ARN, logger logrus.FieldLogger) error {
+	client := ec2.New(session)
+
+	segments := strings.SplitN(arn.Resource, "/", 2)
```

Comment:
from https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arns-syntax  Title `Amazon Elastic Compute Cloud (Amazon EC2)`

`arn:aws:ec2:us-east-1:123456789012:instance/*` is a valid ARN ?

Reply:
Also there are quite a few calls of this type. Can we consolidate them to single function? (I saw 6 same call sites)

Reply:
> > `arn:aws:ec2:us-east-1:123456789012:instance/*` is a valid ARN ?
> 
> It is, see `arn:aws:ec2:region:account-id:instance/instance-id` [here](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-ec2). And we handle it [here](https://github.com/openshift/installer/blob/01d1e81c90dd9f9e138dd904870c1d66c394a820/pkg/destroy/aws/aws.go#L409-L410).

How are we handling it ?, we pass `*` as `id` to `deleteEC2Instance` without checking.. Or are you saying that passing `*` as `id` does not break `deleteEC2Instance`?

> > Also there are quite a few calls of this type. Can we consolidate them to single function? (I saw 6 same call sites)
> 
> The `SplitN` call seemed small enough to inline, but I can wrap it if you like.

I don't have strong opinions, only if it is actually helpful..

Reply:
> > > `arn:aws:ec2:us-east-1:123456789012:instance/*` is a valid ARN ?
> >
> > It is, see `arn:aws:ec2:region:account-id:instance/instance-id` here. And we handle it here.
>
> How are we handling it ?, we pass `*` as `id` to `deleteEC2Instance` without checking..

Oh, I thought you were using `/*` as a wildcard "AWS will put an instance ID in here".  I'm pretty sure we would never get that back as a tagged ARN, although you might use wildcard ARNs like that in [setting up IAM policies][1].  Currently (as of e7b4c4e), `deleteEC2Instance` called with `*` as the ID will fail with the Go equivalent of:

```console
$ aws ec2 describe-instances --instance-ids '*'

An error occurred (InvalidInstanceID.Malformed) when calling the DescribeInstances operation: Invalid id: "*"
```

and that seems sane to me.

> > The SplitN call seemed small enough to inline, but I can wrap it if you like.
>
> I don't have strong opinions, only if it is actually helpful..

I've done this with 01d1e81 -> e7b4c4e, which saves 5 lines overall.  I don't feel strongly either way either ;), let me know if you want me to roll that back.

[1]: https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#genref-aws-service-namespaces